typing/ctype.ml

```diff
@@ -4385,20 +4385,23 @@ let match_class_sig_shape ~strict sign1 sign2 =
       else err)
     sign1.csig_vars errors
 
-let rec moregen_clty trace type_pairs env cty1 cty2 =
+let rec moregen_clty ?arrow_index trace type_pairs env cty1 cty2 =
   try
     match cty1, cty2 with
     | Cty_constr (_, _, cty1), _ ->
         moregen_clty true type_pairs env cty1 cty2
     | _, Cty_constr (_, _, cty2) ->
         moregen_clty true type_pairs env cty1 cty2
```

Comment:
What I wanted to achieve here is this:
When `Option.is_some arrow_index` we are inside a function (`Cty_arrow`). `arrow_index` gives us the current parameter index (so when we encounter the initial `Cty_arrow` it should be `1`).
It should not be set when we're currently not iterating through a function parameter list.

I may have done it in a slightly confusing way, since I'm pre-incrementing the value on the recursive call.
It might be more clear to do the following:
After `|Cty_arrow ... ->`:
```ocaml
let arrow_index = Option.value ~default:0 arrow_index in
let arrow_index = arrow_index + 1 in
```
and then the recursive call will be `moregen_clty ~arrow_index false type_pairs env cty1' cty2'`.
This ensures that on entry to the function, `arrow_index` will be the number of consecutive `Cty_arrow` we've seen so far,  and `None` if we're outside an arrow defintion.

Reply:
>  It should not be set when we're currently not iterating through a function parameter list.

But when we start iterating through the expansion of a class type, we may find arrow types again, and the count will restart from 1, giving a confusing report to users.

(Looking at the parser/grammar, it does not seem possible or at least easy to give a name to a class arrow type, only class signatures `object .. end` are supported after `class type foo =`, so maybe this situation does not occur with the current grammar.)

Reply:
I like the new version; it is not just a subjective matter of style (I hope?), I have the impression that the use of an optional argument helped something that is actually a small bug slip through.